File: mobile/ios/PairMobileKit/Sources/PairMobileKit/PairSessionBootstrapper.swift
import Foundation

/// Result of startup session restoration.
public enum PairSessionBootstrapResult<Snapshot: Sendable>: Sendable {
	case missing
	case restored(Snapshot)
	case invalidated
	case offline(Snapshot)
}

/// Coordinates Keychain and remote validation without imposing application models.
public actor PairSessionBootstrapper<Snapshot: Codable & Sendable> {
	private let store: PairKeychainStore<Snapshot>

	/// Initializes the bootstrapper with the app's persistent store.
	public init(store: PairKeychainStore<Snapshot>) {
		self.store = store
	}

	/// Restores the saved snapshot and validates it with a closure provided by the host app.
	public func bootstrap(
		validate: @Sendable (Snapshot) async throws -> Snapshot
	) async -> PairSessionBootstrapResult<Snapshot> {
		guard let snapshot = await store.load() else {
			return .missing
		}

		do {
			let validatedSnapshot = try await validate(snapshot)
			await store.save(validatedSnapshot)

			return .restored(validatedSnapshot)
		} catch {
			if let apiError = error as? PairAPIError, apiError.isAuthenticationFailure {
				await store.clear()
				return .invalidated
			}

			return .offline(snapshot)
		}
	}
}

File: mobile/android/PairMobileAndroid/src/main/kotlin/dev/pair/mobile/android/PairSessionBootstrapper.kt
package dev.pair.mobile.android

/** Result of startup session restoration. */
sealed class PairSessionBootstrapResult<out Snapshot> {

    /** No saved session exists. */
    data object Missing : PairSessionBootstrapResult<Nothing>()

    /** The saved session was validated and refreshed. */
    data class Restored<Snapshot>(val snapshot: Snapshot) : PairSessionBootstrapResult<Snapshot>()

    /** The backend rejected the saved session. */
    data object Invalidated : PairSessionBootstrapResult<Nothing>()

    /** The saved session is available locally, but remote validation could not complete. */
    data class Offline<Snapshot>(val snapshot: Snapshot) : PairSessionBootstrapResult<Snapshot>()
}

/** Coordinates local storage and remote validation without imposing application models. */
class PairSessionBootstrapper<Snapshot>(
    private val store: PairSessionStore<Snapshot>
) {

    /** Restores the saved snapshot and validates it with a closure provided by the host app. */
    suspend fun bootstrap(
        validate: suspend (Snapshot) -> Snapshot
    ): PairSessionBootstrapResult<Snapshot> {
        val snapshot = store.load() ?: return PairSessionBootstrapResult.Missing

        return try {
            val validatedSnapshot = validate(snapshot)
            store.save(validatedSnapshot)
            PairSessionBootstrapResult.Restored(validatedSnapshot)
        } catch (error: PairApiException) {
            if (error.isAuthenticationFailure) {
                store.clear()
                PairSessionBootstrapResult.Invalidated
            } else {
                PairSessionBootstrapResult.Offline(snapshot)
            }
        } catch (_: Throwable) {
            PairSessionBootstrapResult.Offline(snapshot)
        }
    }
}

